daemon/graphdriver/zfs/zfs.go

```diff
@@ -0,0 +1,277 @@
+package zfs
+
+/*
+#include <locale.h>
+#include <stdlib.h>
+#include <dirent.h>
+#include <mntent.h>
+*/
+import "C"
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+	"unsafe"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/docker/docker/daemon/graphdriver"
+	"github.com/docker/docker/pkg/parsers"
+	zfs "github.com/mistifyio/go-zfs"
+)
+
+type ZfsOptions struct {
+	fsName    string
+	mountPath string
+}
+
+func init() {
+	graphdriver.Register("zfs", Init)
+}
+
+type Logger struct{}
+
+func (*Logger) Log(cmd []string) {
+	log.Debugf("[zfs] %s", strings.Join(cmd, " "))
+}
+
+func Init(base string, opt []string) (graphdriver.Driver, error) {
+	var err error
+	options, err := parseOptions(opt)
+	options.mountPath = base
+	if err != nil {
+		return nil, err
+	}
+
+	rootdir := path.Dir(base)
+
+	if options.fsName == "" {
+		err = checkRootdirFs(rootdir)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if _, err := exec.LookPath("zfs"); err != nil {
+		return nil, fmt.Errorf("zfs command is not available")
+	}
+
+	file, err := os.OpenFile("/dev/zfs", os.O_RDWR, 600)
+	defer file.Close()
```

Comment:
According to close(2): "It is quite possible that errors on a previous write(2) operation are first reported at the final close()" and "...In particular close() should not be retried after an EINTR since this  may cause a reused descriptor from another thread to be close..." 

As there  is no write operation on this file descriptor, there is no benefit in checking for errors on this file descriptor.

btw.
grep -r "defer.*Close()" /usr/lib/go | grep -v "_test" | wc -l